lib/config/config-file.js

```diff
@@ -547,27 +543,60 @@ function load(filePath, configContext, applyEnvironments, relativeTo) {
         }
 
         // validate the configuration before continuing
-        validator.validate(config, filePath, configContext.linterContext.rules, configContext.linterContext.environments);
+        validator.validate(config, resolvedPath, configContext.linterContext.rules, configContext.linterContext.environments);
 
         /*
          * If an `extends` property is defined, it represents a configuration file to use as
          * a "parent". Load the referenced file and merge the configuration recursively.
          */
         if (config.extends) {
-            config = applyExtends(config, configContext, filePath, dirname);
+            config = applyExtends(config, configContext, resolvedPath.filePath, dirname);
         }
+    }
 
-        if (config.env && applyEnvironments) {
+    return config;
+}
 
-            // Merge in environment-specific globals and parserOptions.
-            config = ConfigOps.applyEnvironments(config, configContext.linterContext.environments);
-        }
+/**
+ * Loads a config object, applying extends if present.
+ * @param {Object} configObject a config object to load
+ * @returns {Object} the config object with extends applied if present, or the passed config if not
+ * @private
+ */
+function loadObject(configObject) {
+    return configObject.extends ? applyExtends(configObject, "") : configObject;
+}
+
+/**
+ * Loads a config object from the config cache based on its filename, falling back to the disk if the file is not yet
+ * cached.
+ * @param {string} filePath the path to the config file
+ * @param {Config} configContext Context for the config instance
+ * @param {string} [relativeTo] The path to resolve relative to.
+ * @returns {Object} the parsed config object (empty object if there was a parse error)
+ * @private
+ */
+function load(filePath, configContext, relativeTo) {
+    const resolvedPath = resolve(filePath, relativeTo);
+
+    const cachedConfig = configContext.configCache.getConfig(resolvedPath.filePath);
 
+    if (cachedConfig) {
+        return cachedConfig;
+    }
+
+    const config = loadFromDisk(resolvedPath, configContext);
+
+    if (config) {
+        config.filePath = resolvedPath.filePath;
```

Comment:
Will adding properties to the config object affect the result of `eslint --print-config someFile.js`? Ideally, it would be best to avoid exposing internal properties to the user.

One way to do this could be to have a map of filePath -> config rather than adding a property.